File: scribly/emails.py
from typing import List

from jinja2 import Environment, FileSystemLoader
from premailer import Premailer

from scribly import env
from scribly.definitions import Email, Story, Turn, User
from scribly.jinja_helpers import RemoveNewlines
from scribly.util import read_once

website_url = env.WEBSITE_URL
premailer = Premailer(
    css_text=read_once("static/style.css"), cssutils_logging_level="CRITICAL"
)
jinja_env = Environment(
    loader=FileSystemLoader("email_templates"), extensions=[RemoveNewlines]
)
jinja_env.globals["website_url"] = website_url


def build_added_to_story_emails(story: Story) -> List[Email]:
    assert story.cowriters

    recipients = [
        user
        for user in story.cowriters
        if user.email_verification_status == "verified"
        and not (user.id == story.created_by.id)
    ]

    return [_build_added_to_story_email(story, recipient) for recipient in recipients]


def _build_added_to_story_email(story: Story, recipient: User) -> Email:
    body = _render_template_with_css(
        "addedtostory.html", story=story, recipient=recipient
    )
    subject = f"{story.created_by.username} started the story {story.title}"
    if story.current_writers_turn.id == recipient.id:
        subject += " - it's your turn!"
    return Email(subject=subject, body=body, to=recipient)


def build_turn_email_notifications(story: Story, turn_number: int) -> List[Email]:
    assert len(story.turns) >= turn_number
    assert story.cowriters

    turn = story.turns[turn_number - 1]
    recipients = [
        user
        for user in story.cowriters
        if (user.email_verification_status == "verified")
        and (not user.id == turn.taken_by.id)
    ]

    return [
        _build_turn_email_notification(story, turn, recipient)
        for recipient in recipients
    ]


def _build_turn_email_notification(story: Story, turn: Turn, recipient: User) -> Email:
    body = _render_template_with_css(
        "storyturnnotification.html", story=story, turn=turn, recipient=recipient
    )
    subject: str
    if turn.action in ("pass", "write"):
        assert story.current_writers_turn
        if story.current_writers_turn.id == recipient.id:
            subject = f"It's your turn on {story.title}!"
        else:
            subject = f"{turn.taken_by.username} took their turn on {story.title}!"
    if turn.action in ("write_and_finish", "finish"):
        subject = f"{story.title} is done!"

    return Email(subject=subject, body=body, to=recipient)


def build_email_verification_email(user: User, token: str) -> Email:
    verification_link = f"{website_url}/email-verification?token={token}"
    body = _render_template_with_css(
        "verification.html", verification_link=verification_link, user=user
    )
    return Email(subject="Verify your email", body=body, to=user)


def build_nudge_email(nudger: User, nudgee: User, story: Story) -> Email:
    subject = f"{nudger.username} nudged you to take your turn on {story.title}"
    body = _render_template_with_css(
        "nudge.html", nudger=nudger, nudgee=nudgee, story=story
    )
    return Email(subject, body, to=nudgee)


def _render_template_with_css(template_name: str, **template_kwargs):
    """
    Render an email template and inline styles from stylesheet.
    """
    template = jinja_env.get_template(template_name)
    rendered_html = template.render(**template_kwargs)
    return premailer.transform(rendered_html)

File: internal/emails.go
package internal

import (
	"bytes"
	"fmt"
	"html/template"
	"scribly/embed"
	"strings"

	"github.com/vanng822/go-premailer/premailer"
)

type viewData struct {
	Data       interface{}
	websiteURL string
}

type viewDataOption func(*viewData)

func newViewData(websiteURL string, opts ...viewDataOption) viewData {
	vd := viewData{websiteURL: websiteURL}
	for _, opt := range opts {
		opt(&vd)
	}
	return vd
}

func withData(data interface{}) viewDataOption {
	return func(vd *viewData) {
		vd.Data = data
	}
}

func (v viewData) StoryLink(story Story) template.HTML {
	return template.HTML(fmt.Sprintf(`<a href="%s/stories/%d">%s</a>`, v.websiteURL, story.ID, story.Title))
}

func (v viewData) WebsiteURL() string {
	return v.websiteURL
}

func (v viewData) Replace(original string, pattern string, replacement string) string {
	return strings.ReplaceAll(original, pattern, replacement)
}

func (v viewData) NewLineify(str string) template.HTML {
	return template.HTML(strings.ReplaceAll(str, "\n", "<br>"))
}

func (v viewData) WhoseTurnText(story Story, recipient User) string {
	subject := story.CurrentWriter.Username + "'s"
	if story.CurrentWriterID == recipient.ID {
		subject = "your"
	}
	return fmt.Sprintf("it's %s turn", subject)
}

func (v viewData) CSS() template.HTML {
	return template.HTML("<style>" + embed.CSS + "</style>")
}

func BuildNudgeEmail(websiteURL string, nudger User, nudgee User, story Story) (*Email, error) {
	subject := fmt.Sprintf("%s nudged you to take your turn on %s", nudger.Username, story.Title)
	data := map[string]interface{}{
		"Story":  story,
		"Nudger": nudger,
	}
	body, err := renderTemplateWithCSS("nudge.tmpl", websiteURL, data)
	if err != nil {
		return nil, err
	}

	return &Email{
		Subject: subject,
		Body:    body,
		To:      nudgee,
	}, nil
}

func BuildAddedToStoryEmails(websiteURL string, story Story) ([]Email, error) {
	var recipients []User
	for _, cowriter := range story.Cowriters {
		if cowriter.User.ID != story.CreatedByID && cowriter.User.EmailVerificationStatus == EmailVerificationStateVerified {
			recipients = append(recipients, cowriter.User)
		}
	}

	var emails []Email
	for _, recipient := range recipients {
		data := map[string]interface{}{
			"Story":     story,
			"Recipient": recipient,
		}
		body, err := renderTemplateWithCSS("addedtostory.tmpl", websiteURL, data)
		if err != nil {
			return nil, err
		}
		subject := fmt.Sprintf("%s started the story %s", story.CreatedByU.Username, story.Title)
		if story.CurrentWriterID == recipient.ID {
			subject = subject + " - it's your turn!"
		}
		emails = append(emails, Email{Body: body, Subject: subject, To: recipient})
	}

	return emails, nil
}

func BuildTurnNotificationEmails(websiteURL string, story Story, turnNumber int) ([]Email, error) {
	turn := story.Turns[turnNumber-1]
	var recipients []User
	for _, cowriter := range story.Cowriters {
		if cowriter.User.ID != turn.TakenByID && cowriter.User.EmailVerificationStatus == EmailVerificationStateVerified {
			recipients = append(recipients, cowriter.User)
		}
	}

	var emails []Email
	for _, recipient := range recipients {
		data := map[string]interface{}{
			"Story":      story,
			"TurnNumber": turnNumber,
			"Turn":       &turn,
			"Recipient":  recipient,
		}
		body, err := renderTemplateWithCSS("storyturnnotification.tmpl", websiteURL, data)
		if err != nil {
			return nil, err
		}

		subject := ""
		if turn.Finishes() {
			subject = fmt.Sprintf("%s is done!", story.Title)
		} else {
			if story.CurrentWriterID == recipient.ID {
				subject = fmt.Sprintf("It's your turn on %s!", story.Title)
			} else {
				subject = fmt.Sprintf("%s took their turn on %s!", turn.TakenByU.Username, story.Title)
			}
		}

		emails = append(emails, Email{
			Subject: subject,
			Body:    body,
			To:      recipient,
		})
	}

	return emails, nil
}

func BuildEmailVerificationEmail(websiteURL string, user User, token string) (*Email, error) {
	data := map[string]interface{}{
		"Recipient":         user,
		"VerificationToken": token,
	}
	body, err := renderTemplateWithCSS("verification.tmpl", websiteURL, data)
	if err != nil {
		return nil, err
	}

	return &Email{
		To:      user,
		Subject: "Verify your email",
		Body:    body,
	}, nil
}

func renderTemplateWithCSS(templateName string, websiteURL string, data interface{}) (string, error) {
	var buffer bytes.Buffer
	vd := newViewData(websiteURL, withData(data))
	if err := embed.EmailTemplates.ExecuteTemplate(&buffer, templateName, vd); err != nil {
		return "", err
	}
	prem, err := premailer.NewPremailerFromBytes(buffer.Bytes(), premailer.NewOptions())
	if err != nil {
		return "", err
	}

	return prem.Transform()
}
